cyfronet-fid/marketplace
# [#1941] Ordering API email notifications + refactor

Move logic for email sending on status and voucher_id changes to
ProjectItem::OnStatusTypeUpdated, which is in turn activated
after_commit in ProjectItem.
A similar refactoring takes place for Messages, although there logic is
kept in the Message model.

For now, leave the email sending logic in ProjectItem::Create and
ProjectItem::Ready though, as it seems to be more than just
status-dependent.

Rename WebhookJiraMailer to MessageMailer, as it will no longer only be
used for webhook handling.

Eventable now only creates update Events for allowed attributes from
the parent eventable_attributes.

Tie ProjectItem.statuses creation to after_save to handle update calls
which change both the status and user_secrets in a single transaction.

Similarly, handle user_secrets["voucher_id"] changes in after_save,
creating :user_direct messages as needed.

# How to test

API-key: `XMhD_exLaRJtUMXKsuVn`
oms_id: `1`

Below are services/offers to go through and verify emails are sent (Chipster is locally modified, compared to the initial DB PR content, it won't work the same on master or other PRs).
For each service disregard the values in parens (it's for @jswk not to go crazy).
For each service, as an ordinary user go and order an offer from it (for voucherable don't provide a voucher).
Then set the status_type of such project_item through API to values in list and see if the emails appear, for example for DEEP, set the status_type to "ready", and then you should receive the mail with content [from](https://github.com/cyfronet-fid/marketplace/blob/1941-ordering-api-email-notifications/app/views/project_item_mailer/ready_to_use.html.haml).

To set the status_type of project_item: `PATCH /projects/{project_id}/project_item/{project_item_id}` with payload `{"status":{"value": "whatever", "status_type": "<status_type>"}}`.

To set the voucher_id for a project_item: `PATCH /projects/{project_id}/project_item/{project_item_id}` with payload `{"user_secrets":{"voucher_id": "value"}}`.

DEEP training facility: (order_type=order_required, order_url="")
 - waiting_for_response: :waiting_for_response
 - approved: :approved
 - ready: :ready_to_use
 - rejected: :rejected
 - closed: :closed

100: (order_type=order_required, order_url!="")
 - rejected: :rejected
 - closed: :closed

Chipster: (order_type=order_required, order_url="", aod, non-voucherable)
 - waiting_for_response: :waiting_for_response
 - approved: :approved
 - ready: :aod_accepted
 - rejected: :rejected
 - closed: :closed

EC3 non-voucherable offer: (order_type=order_required, order_url!="", aod, non-voucherable)
 - rejected: :rejected
 - closed: :closed

EC3 voucherable offer: (order_type=order_required, order_url!="", aod, voucherable)
 - set voucher, (message appears: "Voucher has been granted to you...", [email](https://github.com/cyfronet-fid/marketplace/blob/1941-ordering-api-email-notifications/app/views/message_mailer/project_item_new_message.html.haml) is sent)
 - update voucher, (message appears: "Voucher ID has been updated:...", [email](https://github.com/cyfronet-fid/marketplace/blob/1941-ordering-api-email-notifications/app/views/message_mailer/project_item_new_message.html.haml) is sent)
 - set voucher to "" i.e. empty string, (message appears: "Voucher has been revoked", [email](https://github.com/cyfronet-fid/marketplace/blob/1941-ordering-api-email-notifications/app/views/message_mailer/project_item_new_message.html.haml) is sent)
 - rejected: :aod_voucher_rejected
 - closed: :closed

Apart from the status_type emails, there should be emails sent on message creation/update. It shouldn't depend on service type, so can be done with any project and project_item.

Base for message payload:
```json
{
  "scope": "<scope>",
  "content": "<content>",
  "project_id": "<p_id>",
  "project_item_id": "<pi_id>",
  "author": {
    "email": "sdf@sdf.pl",
    "name": "any",
    "role": "provider"
  }
}
```
To create: `POST /oms/{oms_id}/messages`.
To update: `PATCH /oms/{oms_id}/messages/{m_id}` with only `{ "content" : "<content>" } in payload.

1. Create and modify messages with these scopes and see if emails come:
  - scope=public
  - scope=user_direct
 2. Create and modify message with scope=internal and verify the emails don't come.

## How to prepare PR instance

```ruby
Service.find_by(name: "Chipster").update!(order_url: "")
OrderingApi::AddSombo.new.call
Oms.first.id
User.find_by(uid: "iamasomboadmin").authentication_token
```

Closes: #1941.
